aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/ConflictResolver.java

```diff
@@ -0,0 +1,129 @@
+ * local data up to AppSync.
+ *
+ * ConflictResolver's primary entry point is the
+ * {@link ConflictResolver#resolve(PendingMutation, AppSyncConflictUnhandledError)}.
+ * This method will try to rectify state with AppSync by applying the conflict handler
+ * in the {@link DataStoreConfigurationProvider#getConfiguration()}.
+ *
+ * The ConflictResolver will return a {@link ModelWithMetadata} that is understood
+ * to be the final, resolved version of the data, to which both the server and local
+ * database should agree.
+ *
+ * After {@link ConflictResolver#resolve(PendingMutation, AppSyncConflictUnhandledError)}
+ * is invoked by the MutationProcessor, the MutationProcessor must merge the returned
+ * {@link ModelWithMetadata} into the local store, unconditionally.
+ */
+final class ConflictResolver {
+    private final DataStoreConfigurationProvider configurationProvider;
+    private final AppSync appSync;
+
+    ConflictResolver(DataStoreConfigurationProvider configurationProvider, AppSync appSync) {
+        this.configurationProvider = configurationProvider;
+        this.appSync = appSync;
+    }
+
+    @NonNull
+    <T extends Model> Single<ModelWithMetadata<T>> resolve(
+            @NonNull PendingMutation<T> pendingMutation,
+            @NonNull AppSyncConflictUnhandledError<T> conflictUnhandledError) {
+        final DataStoreConflictHandler conflictHandler;
+        try {
+            DataStoreConfiguration configuration = configurationProvider.getConfiguration();
+            conflictHandler = configuration.getDataStoreConflictHandler();
+        } catch (DataStoreException badConfigurationProvider) {
+            return Single.error(badConfigurationProvider);
+        }
+
+        ModelWithMetadata<T> serverData = conflictUnhandledError.getServerVersion();
+        ModelMetadata metadata = serverData.getSyncMetadata();
+        T remote = serverData.getModel();
+        T local = pendingMutation.getMutatedItem();
+        ConflictData<T> conflictData = ConflictData.create(local, remote);
+
+        return Single
+            .<ConflictResolutionDecision<? extends Model>>create(emitter ->
+                conflictHandler.onConflictDetected(conflictData, emitter::onSuccess)
+            )
+            .flatMap(decision -> {
+                @SuppressWarnings("unchecked")
+                ConflictResolutionDecision<T> typedDecision = (ConflictResolutionDecision<T>) decision;
+                return resolveModelAndMetadata(conflictData, metadata, typedDecision);
+            });
+    }
+
+    @NonNull
+    private <T extends Model> Single<ModelWithMetadata<T>> resolveModelAndMetadata(
+            @NonNull ConflictData<T> conflictData,
+            @NonNull ModelMetadata metadata,
+            @NonNull ConflictResolutionDecision<T> decision) {
+        // If the strategy is apply remote or retry local, we already know the version we'll use.
+        // We simply extract it from the conflict data.
```

Comment:
It looks like this is done in all three cases below - not understanding the comment.

Reply:
Good call! This comment looks like a stray. I'll remote it.